internal/rpmver/rpmver.go

```diff
@@ -0,0 +1,338 @@
+func (v *Version) EVR() string {
+	var b strings.Builder
+	v.evr(&b)
+	return b.String()
+}
+
+// Parse returns a Version for the provided string, or an error if it's
+// malformed.
+func Parse(v string) (Version, error) {
+	ret := Version{
+		Epoch: "0",
+	}
+	switch strings.Count(v, "-") {
+	case 0:
+		// Missing something: can't be `version-release`.
+		return Version{}, fmt.Errorf("rpmver: %s: missing separators", v)
+	case 1:
+		// `version-release(.arch)`
+	default:
+		// `some-name-version-release(.arch)`
+		i := strings.LastIndexByte(v, '-')
+		i = strings.LastIndexByte(v[:i], '-')
+		// Can't be -1, there are at least two "-".
+		name := v[:i]
+		ret.Name = &name
+		v = v[i+1:]
+	}
+	ev, ra, _ := strings.Cut(v, "-")
+
+	ret.Version = ev
+	if e, v, ok := strings.Cut(ev, ":"); ok {
+		if e != "" {
+			ret.Epoch = e
+		}
+		ret.Version = v
+	}
+
+	ret.Release = ra
+	if idx := strings.LastIndexByte(ra, '.'); idx != -1 {
+		a := ra[idx:]
+		if _, ok := architectures[a]; ok {
+			arch := a[1:]
+			ret.Architecture = &arch
+			ret.Release = ra[:idx]
+		}
+	}
+
+	return ret, nil
+}
+
+// Architectures is known architecture strings.
+//
+// We need to just know these, as there's no good way to know what's an arch tag
+// and what's just another version segment.
+var architectures = map[string]struct{}{
+	".aarch64": {},
+	".i686":    {},
+	".noarch":  {},
+	".ppc64le": {},
+	".riscv":   {},
+	".s390x":   {},
+	".src":     {},
+	".x86_64":  {},
+}
+
+// Cmp is a mnemonic helper for the comparison result type.
+//
+// This can't be the return type for [Compare] because all the users expect an
+// [int] return.
+type cmp int
+
+//go:generate go run golang.org/x/tools/cmd/stringer -type cmp -linecomment -output cmp_string_test.go
+
+const (
+	cmpLT cmp = iota - 1 // <
+	cmpEQ                // ==
+	cmpGT                // >
+)
+
+// Compare is a comparison for Versions.
+func Compare(a, b *Version) int {
+	if cmp := comparePtr(a.Name, b.Name); cmp != 0 {
+		return cmp
+	}
+
+	if cmp := rpmvercmp(a.Epoch, b.Epoch); cmp != 0 {
+		return cmp
+	}
+
+	if cmp := rpmvercmp(a.Version, b.Version); cmp != 0 {
+		return cmp
+	}
+
+	if cmp := rpmvercmp(a.Release, b.Release); cmp != 0 {
+		return cmp
+	}
+
+	if cmp := comparePtr(a.Architecture, b.Architecture); cmp != 0 {
+		return cmp
+	}
+
+	return int(cmpEQ)
+}
+
+// ComparePtr runs [rpmvercmp] after considering the pointer-ness of the values.
+func comparePtr(a, b *string) int {
+	switch {
+	case a == nil && b == nil:
+		return int(cmpEQ)
+	case a != nil && b == nil:
+		return int(cmpGT)
+	case a == nil && b != nil:
+		return int(cmpLT)
+	default:
+	}
+	return rpmvercmp(*a, *b)
+}
+
+// Rpmvercmp compares RPM version strings.
+//
+// This is a port of the C version at https://github.com/rpm-software-management/rpm/blob/572844039a04846fe9e030cbacb6336e2240bd6f/rpmio/rpmvercmp.cc
+//
+//	 1: a is newer than b
+//	 0: a and b are the same version
+//	-1: b is newer than a
+func rpmvercmp(a, b string) int {
```

Comment:
It's hard to tell, but I don't think this has changed since I last reviewed. When I last reviewed, this all looked good to me after I did a comparison between the C version and this version